[Feature] Allow embargo registrations at 3 days out [OSF-4304]

mluke93 opened the pull request:
<!-- Before submit your Pull Request, make sure you picked
     the right branch:

     - For hotfixes, select "master" as the target branch
     - For new features, select "develop" as the target branch
     - For release feature fixes, select the relevant release branch (release/X.Y.Z) as the target branch -->

## Purpose

When attempting to embargo a Registration, if the user attempts to pick a date 3 days away (as is required to Embargo) after 8pm, the system informs them that they must pick a date at least 3 days away, and won't let them pick that date (ie. On August 31st 8:36pm, the calendar does not recognize September 3rd as a valid date). The problem lied in the timezone difference between local time and UTC time. There is a four hour difference (hence why it had a problem starting at 8pm). So the valid date calculation needed to account for the time zone difference.

## Changes

- Added the Timezone Offset when calculating the chosen end date in milliseconds
- Fixed some of formatting.

## Side effects

- The max length (4 years) might have a similar issue to this.

## Ticket

https://openscience.atlassian.net/browse/OSF-4304

brianjgeiger commented on website/static/js/registrationModal.js:
```diff
@@ -52,14 +52,15 @@ var RegistrationViewModel = function(confirm, prompts, validator) {
     var validation = [{
         validator: function() {
-            var endEmbargoDateTimestamp = self.embargoEndDate().getTime();
+            var timeZoneOffset = self.embargoEndDate().getTimezoneOffset() * (60 * 1000);
```
We should still be using moment.js in the site. There's probably a better way to do this using moment. The things I want to avoid are 1) using magic numbers; and 2) anything by hand related to time because time is the *worst*.

brianjgeiger commented on website/static/js/registrationModal.js:
```diff
@@ -52,14 +52,15 @@ var RegistrationViewModel = function(confirm, prompts, validator) {
 	validator: function() {
             var endEmbargoDateTimestamp = self.embargoEndDate().getTime();
-	    return endEmbargoDateTimestamp < FOUR_YEARS_FROM_TODAY_TIMESTAMP;
+	        return endEmbargoDateTimestamp < FOUR_YEARS_FROM_TODAY_TIMESTAMP;
```
Is this too many extra spaces?

mluke93 replied:
It looked wrong on my end, so I changed something. Let me go back and fix that, because you're definitely right. That doesn't look right at all.

brianjgeiger commented:
Pass done. Mostly just look into http://momentjs.com

mluke93 commented:
Made changes to the validation so it now uses a moment to calculate the time chosen on the calendar (including UTC offsets). Also, made the same change for the 'Four Years' case, so it is consistent (before, you could sometimes pick exactly four years out, and sometimes you couldn't). You now must pick a date under 4 years out.

brianjgeiger commented on website/static/js/registrationModal.js:
```diff
@@ -52,14 +53,14 @@ var RegistrationViewModel = function(confirm, prompts, validator) {
 	validator: function() {
-            var endEmbargoDateTimestamp = self.embargoEndDate().getTime();
-	    return endEmbargoDateTimestamp < FOUR_YEARS_FROM_TODAY_TIMESTAMP;
+            var endEmbargoMoment = moment(self.embargoEndDate()).subtract(moment().utcOffset(), 'm');
```
Spacing in this function block seems all wrong. Four spaces to indent. https://cosdev.readthedocs.io/en/latest/style_guides/javascript.html#style

brianjgeiger commented:
Okay, *since* we're in here already and I'm having a hard time figuring out for sure what's going on, maybe it's worth doing just a bit more to make it better. Right now, limits are being set at https://github.com/CenterForOpenScience/osf.io/pull/5708/files#diff-9f8b0649d71a07ff6a3d91a6348fcb92R16 which looks like it's going to happen at instantiation rather than at validation, so if you load the page, wait three days, and then try to embargo for 30 minutes from now, it'll work, because you'll be past the three day wait time on account of waiting. Not the worst thing, but there'll be some other inconsistencies and there's no reason for it.

More importantly, it's really hard to tell from what's happening here what's in local time and what's in UTC time at which point, and so maybe we can make this a little bit better now that we have moment. Finally, there are no tests, which makes it especially difficult for me to be sure it's fixed properly. So add some story points to the issue and let's see if we can do this correctly.

Moment can add dates together by doing super readable things like `moment().add(7, 'days').add(1, 'months');` so we can take advantage of that. If we name the validator functions instead of declaring them anonymously in the validator, then we can write some unit tests that verify what will happen in different conditions.

So essentially I'm asking you to make:

- minimumTimeValidation(embargoLocalDateTime) function
- maximumTimeValidation(embargoLocalDateTime) function
- Unit tests for the two validation functions above to test normal functioning, plus some likely problems when people are using the browser in different time zones. (You can pass in the embargoLocalDateTime in whichever timezone that the person requests, that way you're not relying on someone upstream to convert from localtime - you want all the conversion to happen behind the tested function.)

Does this all make sense? It might not work exactly like I'm suggesting, but the important thing is to make it more readable and to make it tested.

brianjgeiger commented:
Pass done. 🐧

brianjgeiger commented:
You might also consider having separate functions to set the minimum time and the maximum time outside of the validator functions, and have your tests use that as well, so if we change the minimum time and maximum time in the future, the tests won't have to be adjusted. Since there's a todo near the code that sets the min/max times to improve that, we know it has a chance of being changed, so let's be prepared.

mluke93 commented:
I think I see what you're going for. I will definitely go ahead and take a
crack at this. If I have any further questions I'll drop by and ask you.

On Tue, May 24, 2016 at 9:07 AM, Brian J. Geiger <notifications@github.com>
wrote:

> You might also consider having separate functions to set the minimum time
> and the maximum time outside of the validator functions, and have your
> tests use that as well, so if we change the minimum time and maximum time
> in the future, the tests won't have to be adjusted. Since there's a todo
> near the code that sets the min/max times to improve that, we know it has a
> chance of being changed, so let's be prepared.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/CenterForOpenScience/osf.io/pull/5708#issuecomment-221262494>
>



-- 
Luke Marsh

brianjgeiger commented:
So you know, I'm working from home today but will be in tomorrow. However, there are folks who are ready to mentor, so just ask on osf-dev for them.

mluke93 commented:
Good to know! Thanks for the advice.

mluke93 commented:
Let me know if something isn't formatted correctly, or if you still want anything changed. The validation functions are inside of the registrationViewModel, while the getMin/MaxTime are outside. If you have a preference on location of these functions, I can go ahead and change that. Also added 6 tests (2 regular valid, 2 regular invalid, 2 timezone-based). The last two should work in whatever timezone the computer is in.

brianjgeiger commented on website/static/js/registrationModal.js:
```diff
@@ -47,21 +45,31 @@ var RegistrationViewModel = function(confirm, prompts, validator) {
-        validator: function() {
-            var endEmbargoDateTimestamp = self.embargoEndDate().getTime();
-            return endEmbargoDateTimestamp > TWO_DAYS_FROM_TODAY_TIMESTAMP;
+        validator: function () {
```
I would expect this could just be `validator: self.minimumTimeValidation,` or similar.

brianjgeiger commented on website/static/js/registrationModal.js:
```diff
@@ -47,21 +45,31 @@ var RegistrationViewModel = function(confirm, prompts, validator) {
-	    return endEmbargoDateTimestamp < FOUR_YEARS_FROM_TODAY_TIMESTAMP;
-	},
-	message: 'Embargo end date must be less than four years in the future.'
+        validator: function () {
```
And here. The advantage of this is that, when you're debugging, you don't have an anonymous function in your call stack. It makes things much easier.

brianjgeiger commented on website/static/js/registrationModal.js:
```diff
@@ -103,10 +111,20 @@ RegistrationViewModel.prototype.show = function() {
     ViewModel: RegistrationViewModel
 };
+
+function getMinimumDate(embargoLocalDateTime) {
```
👍

brianjgeiger commented on website/static/js/tests/registrationModal.test.js:
```diff
@@ -104,4 +107,44 @@ describe('registrationModal', () => {
+        it('returns true for date less than 4 years in the future, regardless of time zone', () => {
+            var validDateFourYearsTZ = new Date(2016, 0, 1);
+            validDateFourYearsTZ.setMinutes(validDateFourYearsTZ.getMinutes() - validDateFourYearsTZ.getTimezoneOffset());
+            console.log(validDateFourYearsTZ);
```
No console logs in PRs, please.

brianjgeiger commented on website/static/js/tests/registrationModal.test.js:
```diff
@@ -104,4 +107,44 @@ describe('registrationModal', () => {
+            console.log(validDateFourYearsTZ);
+            var validDateTZFutureYears = new Date(validDateFourYearsTZ);
+            validDateTZFutureYears.setDate(validDateTZFutureYears.getDate() + 1460);
+            console.log(validDateTZFutureYears);
```
Same here

brianjgeiger commented:
Tests are mostly there. Do the false versions of the tests with timezones, and do a couple variants of timezones (such as one that's west of UTC and one that's east).

brianjgeiger commented:
Overall, this is looking much more maintainable and comprehensible. Very good. Pass done. 🐧

mluke93 commented:
Whoops! Sorry about the console logs. That was an oversight on my part. I used them for my own personal testing. I'll fix up those validator functions and add in the extra tests.

brianjgeiger commented:
Heh. Happens all the time. There's one developer who has to put money in a jar each time. Since this is your first offense, we won't do that to you. :)

mluke93 commented:
Alright, let's have another go. Sorry about the delay, been working on our project. So, I've removed the anonymous function, but did have to add in the parameter workaround on minimumTimeValidation and maximumTimeValidation. From everything I've tried, it seems we either get an anonymous function or those extra two parameters we don't need. The tests have been updated and should reflect varying time zones. I believe that's all I've changed.

brianjgeiger commented:
Looks like you have some problems with tests. At the very least a few JSHint errors. 

![screen shot 2016-06-01 at 2 49 47 pm](https://cloud.githubusercontent.com/assets/6599111/15721769/6cc69664-2808-11e6-968c-d911ab465aa7.png)

brianjgeiger commented on website/static/js/registrationModal.js:
```diff
@@ -47,21 +45,29 @@ var RegistrationViewModel = function(confirm, prompts, validator) {
-            return endEmbargoDateTimestamp > TWO_DAYS_FROM_TODAY_TIMESTAMP;
-        },
+        validator: self.minimumTimeValidation
+        ,
```
I imagine this was one of the hint errors, but move this comma up to the previous line.

mluke93 commented:
Oh. Huh. Ran fine on my end. I'l check it out. Sorry about that.

brianjgeiger commented on website/static/js/tests/registrationModal.test.js:
```diff
@@ -104,4 +107,90 @@ describe('registrationModal', () => {
+            vm.pikaday(validDateTZFutureWest);
+            assert.isTrue(vm.maximumTimeValidation(null, null, validDateFourYearsTZWest));
+        });
+        it('returns true for date less than 4 years in the future, in a western timezone', () => {
```
I think this should be `eastern`

brianjgeiger commented on website/static/js/tests/registrationModal.test.js:
```diff
@@ -104,4 +107,90 @@ describe('registrationModal', () => {
+            vm.pikaday(invalidDateTZFuture);
+            assert.isFalse(vm.minimumTimeValidation(null, null, invalidDateTwoDaysTZ));
+        });
+        it('returns false for date at least 4 years in the future, regardless of time zone', () => {
```
I think 'regardless of time zone' might not be the right thing to say here. Can you rephrase, or explain if the phrasing is exactly perfect?

brianjgeiger commented:
Super close. Pass don. 🐧

mluke93 commented:
Hopefully this should do it.

brianjgeiger commented:
If travis passes, this is good to merge. 🎉 🐧 🎉

mluke93 commented:
Woohoo!

brianjgeiger commented:
We have a travis test that runs tests on the code after it's been merged with develop. In this case, it looks like a Javascript test is failing in the post-merge scenario. 

![screen shot 2016-06-01 at 5 50 32 pm](https://cloud.githubusercontent.com/assets/6599111/15726941/a849e97a-2821-11e6-9b20-8fa007b5e37d.png)

If you pull in develop and run Karma tests, you should see it and be able to fix it. Everything else looks good.

mluke93 commented:
Huh. Interesting. Ill get on that first thing in the morning.
On Jun 1, 2016 5:54 PM, "Brian J. Geiger" <notifications@github.com> wrote:

> We have a travis test that runs tests on the code after it's been merged
> with develop. In this case, it looks like a Javascript test is failing in
> the post-merge scenario.
>
> [image: screen shot 2016-06-01 at 5 50 32 pm]
> <https://cloud.githubusercontent.com/assets/6599111/15726941/a849e97a-2821-11e6-9b20-8fa007b5e37d.png>
>
> If you pull in develop and run Karma tests, you should see it and be able
> to fix it. Everything else looks good.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/CenterForOpenScience/osf.io/pull/5708#issuecomment-223136565>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ANGWSj7DifQQg6zmSu9sHuE101cvCcnvks5qHf95gaJpZM4Ijea8>
> .
>

brianjgeiger commented:
This is good. It was a random test failure.

sloria commented on website/static/js/tests/registrationModal.test.js:
```diff
@@ -104,4 +107,58 @@ describe('registrationModal', () => {
         });
     });
+    describe('#timeValidation', () => {
+        it('returns true for date more than 2 days in the future', () => {
```
Thanks for being thorough with testing.
